diff --git a/components/Product.tsx b/components/Product.tsx
index f2ec2d8..2fd0de4 100644
--- a/components/Product.tsx
+++ b/components/Product.tsx
@@ -11,9 +11,12 @@ import Typography from "@mui/material/Typography";
 import { red } from "@mui/material/colors";
 import MoreVertIcon from "@mui/icons-material/MoreVert";
 import AddShoppingCartIcon from "@mui/icons-material/AddShoppingCart";
+import { ShopContext } from "../context/shop-context";
+import { ProductType } from "../products";
 
-export default function Product() {
+export default function Product({ id, name, description, image, amount }: ProductType) {
   const [expanded, setExpanded] = React.useState(false);
+  const { cartItems, addToCart } = React.useContext(ShopContext);
 
   const handleExpandClick = () => {
     setExpanded(!expanded);
@@ -32,24 +35,27 @@ export default function Product() {
             <MoreVertIcon />
           </IconButton>
         }
-        title="Shrimp and Chorizo Paella"
+        title={name}
         subheader="September 14, 2016"
       />
       <CardMedia
         component="img"
         height="194"
-        image="/static/images/cards/paella.jpg"
+        image={image}
         alt="Paella dish"
       />
       <CardContent>
         <Typography variant="body2" color="text.secondary">
-          This impressive paella is a perfect party dish and a fun meal to cook
-          together with your guests. Add 1 cup of frozen peas along with the
-          mussels, if you like.
+          {description}
         </Typography>
+        { amount !== undefined && amount > 0 && 
+          <Typography variant="body2" color="text.secondary">
+            Amount: {amount}
+          </Typography>
+        }
       </CardContent>
       <CardActions disableSpacing>
-        <IconButton aria-label="add to shopping cart">
+        <IconButton aria-label="add to shopping cart" onClick={() => addToCart(id)}>
           <AddShoppingCartIcon />
         </IconButton>
       </CardActions>
diff --git a/components/Products.tsx b/components/Products.tsx
index c54efe0..7109182 100644
--- a/components/Products.tsx
+++ b/components/Products.tsx
@@ -9,6 +9,7 @@ import ListItemAvatar from "@mui/material/ListItemAvatar";
 import Avatar from "@mui/material/Avatar";
 import Typography from "@mui/material/Typography";
 import Product from "./Product";
+import { PRODUCTS } from "../products";
 
 const Products = () => {
   return (
@@ -24,21 +25,18 @@ const Products = () => {
         <List
           sx={{ width: "100%", maxWidth: 360, bgcolor: "background.paper" }}
         >
-          <ListItem alignItems="flex-start">
-            <Product />
-          </ListItem>
-          <Divider variant="inset" component="li" />
-          <ListItem alignItems="flex-start">
-            <Product />
-          </ListItem>
-          <Divider variant="inset" component="li" />
-          <ListItem alignItems="flex-start">
-            <Product />
-          </ListItem>
-          <Divider variant="inset" component="li" />
-          <ListItem alignItems="flex-start">
-            <Product />
-          </ListItem>
+          { PRODUCTS.map((product) => 
+          <div>
+            <ListItem alignItems="flex-start">
+              <Product 
+                id={product.id} 
+                name={product.name} 
+                description={product.description} 
+                image={product.image}
+              />
+            </ListItem>
+            <Divider variant="inset" component="li" />
+          </div>) }
         </List>
       </Box>
     </Container>
diff --git a/components/TopBar.tsx b/components/TopBar.tsx
index e554cfa..5f6e6b2 100644
--- a/components/TopBar.tsx
+++ b/components/TopBar.tsx
@@ -11,12 +11,17 @@ import FormControlLabel from "@mui/material/FormControlLabel";
 import FormGroup from "@mui/material/FormGroup";
 import MenuItem from "@mui/material/MenuItem";
 import Menu from "@mui/material/Menu";
+import Badge from "@mui/material/Badge";
 import ShoppingCartIcon from "@mui/icons-material/ShoppingCart";
 import { useRouter } from "next/router";
+import { ShopContext } from "../context/shop-context";
 
 export default function MenuAppBar() {
   const router = useRouter();
   const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
+  const {cartItems, addToCart} = React.useContext(ShopContext);
+
+  let itemQuantity = Object.values<number>(cartItems).reduce((t, value) => t + value, 0);
 
   const handleMenu = (event: React.MouseEvent<HTMLElement>) => {
     setAnchorEl(event.currentTarget);
@@ -58,7 +63,9 @@ export default function MenuAppBar() {
             onClick={handleCart}
             color="inherit"
           >
-            <ShoppingCartIcon />
+            <Badge badgeContent={itemQuantity} color="secondary">
+              <ShoppingCartIcon />
+            </Badge>
           </IconButton>
           <IconButton
             size="large"
diff --git a/package-lock.json b/package-lock.json
index 3e905cb..dcc1b25 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -4,7 +4,6 @@
   "requires": true,
   "packages": {
     "": {
-      "name": "muxt-ts",
       "dependencies": {
         "@emotion/cache": "^11.7.1",
         "@emotion/react": "^11.9.0",
diff --git a/pages/_app.tsx b/pages/_app.tsx
index 21d5af4..d2217bc 100644
--- a/pages/_app.tsx
+++ b/pages/_app.tsx
@@ -11,6 +11,7 @@ import '@fontsource/roboto/700.css';
 import createEmotionCache from '../utility/createEmotionCache';
 import lightThemeOptions from '../styles/theme/lightThemeOptions';
 import '../styles/globals.css';
+import { ShopContextProvider } from '../context/shop-context';
 interface MyAppProps extends AppProps {
   emotionCache?: EmotionCache;
 }
@@ -25,8 +26,10 @@ const MyApp: React.FunctionComponent<MyAppProps> = (props) => {
   return (
     <CacheProvider value={emotionCache}>
       <ThemeProvider theme={lightTheme}>
-        <CssBaseline />
-        <Component {...pageProps} />
+        <ShopContextProvider value={0}>
+          <CssBaseline />
+          <Component {...pageProps} />
+        </ShopContextProvider>
       </ThemeProvider>
     </CacheProvider>
   );
diff --git a/pages/cart.tsx b/pages/cart.tsx
index ef6ec88..83c3942 100644
--- a/pages/cart.tsx
+++ b/pages/cart.tsx
@@ -3,9 +3,13 @@ import TopBar from "../components/TopBar";
 import { Box, Button, Container, Divider, List, ListItem } from "@mui/material";
 import { useRouter } from "next/router";
 import Product from "../components/Product";
+import { ShopContext } from "../context/shop-context";
+import React from "react";
+import { PRODUCTS } from "../products";
 
 const Cart: NextPage = () => {
   const router = useRouter();
+  const {cartItems, addToCart} = React.useContext(ShopContext);
 
   const handleCheckout = () => {
     router.push("checkout");
@@ -26,10 +30,22 @@ const Cart: NextPage = () => {
           <List
             sx={{ width: "100%", maxWidth: 360, bgcolor: "background.paper" }}
           >
-            <ListItem alignItems="flex-start">
-              <Product />
-            </ListItem>
-            <Divider variant="inset" component="li" />
+          { PRODUCTS.map((product) => {
+            if (cartItems[product.id] > 0) {
+              return <div>
+                <ListItem alignItems="flex-start">
+                  <Product 
+                    id={product.id} 
+                    name={product.name} 
+                    description={product.description} 
+                    image={product.image}
+                    amount={cartItems[product.id]}
+                  />
+                </ListItem>
+                <Divider variant="inset" component="li" />
+              </div>;
+            }
+          })}
           </List>
         </Box>
       </Container>
